### PR TITLE
feat: use ShipEngine for fetching shipping rates

### PR DESCRIPTION
**Pull request by Alchez**
**Ref:** [TASK-2020-00209](https://bloomstack.com/desk#Form/Task/TASK-2020-00209)

### CODE REVIEW

File: shipment_management/public/js/get_rates.js [hunk 1]
```diff
@@ -1,92 +1,109 @@
...
+				"days": null,
+				"arrival": null
+			});
+
+			// display all the shipping options
+			frappe.prompt({
+				"label": "Service Types",
+				"fieldtype": "Select",
+				"fieldname": "service_types",
+				"options": options,
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:87]: Codacy Issue found: [Expected property shorthand.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 2]
```diff
@@ -1,92 +1,109 @@
...
+				let service_data = service_dict.find((rate) => rate.label == selected_service);
+
+				if (service_data.fee != 0) {
+					frm.add_child("taxes", {
+						"charge_type": "Actual",
+						"account_head": "Freight and Forwarding Charges - JA",
+						"tax_amount": service_data.fee,
+						"description": `Shipping (${service_data.label}")`
+					});
+				};
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:101]: Codacy Issue found: [Unnecessary semicolon.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 3]
```diff
@@ -1,92 +1,109 @@
...
+			},
+			function(data) {
+				let selected_service = data.service_types;
+				let service_data = service_dict.find((rate) => rate.label == selected_service);
+
+				if (service_data.fee != 0) {
+					frm.add_child("taxes", {
+						"charge_type": "Actual",
+						"account_head": "Freight and Forwarding Charges - JA",
+						"tax_amount": service_data.fee,
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:98]: Codacy Issue found: [Identifier 'service_data' is not in camel case.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 4]
```diff
@@ -1,92 +1,109 @@
...
+					"label": option,
+					"code": value.code,
+					"fee": value.fee,
+					"days": value.days,
+					"arrival": value.estimated_arrival
+				});
+			});
+
+			options.push("Pick Up - $0");
+			service_dict.push({
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:74]: Codacy Issue found: [Identifier 'service_dict' is not in camel case.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 5]
```diff
@@ -1,92 +1,109 @@
...
+			// display all the shipping options
+			frappe.prompt({
+				"label": "Service Types",
+				"fieldtype": "Select",
+				"fieldname": "service_types",
+				"options": options,
+				"reqd": 1
+			},
+			function(data) {
+				let selected_service = data.service_types;
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:91]: Codacy Issue found: [Identifier 'selected_service' is not in camel case.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 6]
```diff
@@ -1,92 +1,109 @@
...
+	frappe.call({
+		freeze: 1,
+		method: 'shipment_management.shipengine.api.get_rates',
+		args: {
+			doc: frm.doc,
+			estimate: false
+		},
+		callback: function(response) {
+			let options = [];
+			let service_dict = [];
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:56]: Codacy Issue found: [Identifier 'service_dict' is not in camel case.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 7]
```diff
@@ -1,92 +1,109 @@
...
+			frappe.prompt({
+				"label": "Service Types",
+				"fieldtype": "Select",
+				"fieldname": "service_types",
+				"options": options,
+				"reqd": 1
+			},
+			function(data) {
+				let selected_service = data.service_types;
+				let service_data = service_dict.find((rate) => rate.label == selected_service);
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:92]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 8]
```diff
@@ -1,92 +1,109 @@
...
-                    if (service_data.fee != 0) {
-                        let row = frappe.model.add_child(frm.doc, "Sales Taxes and Charges", "taxes");
-                        row.charge_type = "Actual";
-                        row.account_head = "Freight and Forwarding Charges - JA";
-                        row.tax_amount = service_data.fee
-                        row.description = "Shipping(" + service_data.label + ")"
+			// form all the available shipping options
+			$.each(response.message, function(index, value) {
+				let estimated_arrival = isNaN(value.estimated_arrival) ? `(${value.estimated_arrival})` : `(${value.days} days)`;
+				let option = `${value.label} - $${value.fee} ${estimated_arrival}`;
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:61]: Codacy Issue found: [Identifier 'estimated_arrival' is not in camel case.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 9]
```diff
@@ -1,92 +1,109 @@
...
+				if (service_data.fee != 0) {
+					frm.add_child("taxes", {
+						"charge_type": "Actual",
+						"account_head": "Freight and Forwarding Charges - JA",
+						"tax_amount": service_data.fee,
+						"description": `Shipping (${service_data.label}")`
+					});
+				};
+
+				frm.set_value("fedex_shipping_method", service_data.code);
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:103]: Codacy Issue found: [Identifier 'service_data' is not in camel case.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 10]
```diff
@@ -1,92 +1,109 @@
...
+			function(data) {
+				let selected_service = data.service_types;
+				let service_data = service_dict.find((rate) => rate.label == selected_service);
+
+				if (service_data.fee != 0) {
+					frm.add_child("taxes", {
+						"charge_type": "Actual",
+						"account_head": "Freight and Forwarding Charges - JA",
+						"tax_amount": service_data.fee,
+						"description": `Shipping (${service_data.label}")`
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:99]: Codacy Issue found: [Identifier 'service_data' is not in camel case.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js [hunk 11]
```diff
@@ -1,92 +1,109 @@
...
-        method: 'shipment_management.api.get_rates_for_doc',
-        args: {
-            doc: frm.doc
-        },
-        callback: function (response) {
-            let options_string = ""
-            let service_dict = {}
+	frm.doc.taxes.forEach((item) => {
+		if (item.account_head == "Freight and Forwarding Charges - JA") {
+			frappe.throw("Shipment Charge has already been added");
```
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:43]: Codacy Issue found: ['frappe' is not defined.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

File: shipment_management/public/js/get_rates.js (hunk 7 above)
Comment (vjFaLk) [inline shipment_management/public/js/get_rates.js:92]: Codacy Issue found: [Identifier 'selected_service' is not in camel case.](https://app.codacy.com/manual/mail_67/shipment_management/pullRequest?prid=5409514)

**Review by DeveloperMetal — APPROVED** [review]

**Comment by DeveloperMetal** [comment]
Tested all weekend. Functionality seems intact from previous fedex api

**Pull request merged**